several atomic comparisons raise wrong error code

### What is the problem

Depending on the order of an atomic comparison, eXist inconsistently raises either a generic error (`exerr:ERROR`), `err:FORG0001`, or the expected `err:XPTY0004` error.

This appears to be due to inconsistent practice in the implementation of the `compareTo(Collator, Comparison, AtomicValue)` method in each of the `org.exist.xquery.value` classes. Specifically, `AbstractDateTimeValue`, `BinaryValue`, and `NumericValue` all raise generic errors. `AnyURIValue` raises `err:FORG0001` if compared to anything that cannot be cast to an `xs:anyURI`.

### What did you expect

Comparisons of incompatible types should always raise `err:XPTY0004`, regardless of the comparison order and types involved. Eg. `5 eq 'five'`, `'five' eq 5`, and `xs:anyURI('5') eq 5` should all raise `err:XPTY0004`.

### Context information
- eXist-db 3.6.0 (build 201711271542) / ceaab01
- Java 1.8.0_152-b16
- MacOS 10.13.1
- 64 bit

Related: #1630

as I see it, err:XPTY0004  can be thrown into context of sequence matching whilst err:FORG0001 is thrown in the context of type casting. The stacktrace  of your examples indicate this direction:

```
Caused by: org.exist.xquery.XPathException: err:FORG0001 cannot convert 'xs:integer (5)' into xs:anyURI
	at org.exist.xquery.value.IntegerValue.convertTo(IntegerValue.java:317)
	at org.exist.xquery.value.AnyURIValue.compareTo(AnyURIValue.java:297)
``` 

for sure `5 eq 'five'` should yield into a better error....

--- a/src/org/exist/xquery/value/AbstractDateTimeValue.java
+++ b/src/org/exist/xquery/value/AbstractDateTimeValue.java
@@ -359,9 +359,8 @@ public abstract class AbstractDateTimeValue extends ComputableValue {
             }
             return r;
         }
-        throw new XPathException(
-                "Type error: cannot compare " + Type.getTypeName(getType()) + " to "
-                        + Type.getTypeName(other.getType()));
+        throw new XPathException(ErrorCodes.XPTY0004, "Type error: cannot compare " + Type.getTypeName(getType())
+                + " to " + Type.getTypeName(other.getType()));
     }
 
     public AtomicValue max(Collator collator, AtomicValue other) throws XPathException {

--- a/src/org/exist/xquery/value/AnyURIValue.java
+++ b/src/org/exist/xquery/value/AnyURIValue.java
@@ -37,7 +37,7 @@ import java.util.BitSet;
 /**
  * @author Wolfgang Meier (wolfgang@exist-db.org)
  */
-public class AnyURIValue extends AtomicValue {
+public class    AnyURIValue extends AtomicValue {
 
     public static final AnyURIValue EMPTY_URI = new AnyURIValue();
     static final int caseDiff = ('a' - 'A');
@@ -288,13 +288,24 @@ public class AnyURIValue extends AtomicValue {
                 case LTEQ:
                     return cmp <= 0;
                 default:
-                    throw new XPathException(
-                            "XPTY0004: cannot apply operator "
-                                    + operator.generalComparisonSymbol
-                                    + " to xs:anyURI");
+                    throw new XPathException(ErrorCodes.XPTY0004,
+                            "cannot apply operator " + operator.generalComparisonSymbol + " to xs:anyURI");
             }
         } else {
-            return compareTo(collator, operator, other.convertTo(Type.ANY_URI));
+            AtomicValue atomicValue = null;
+
+            try {
+                atomicValue = other.convertTo(Type.ANY_URI);
+
+            } catch (XPathException ex) {
+
+                if (ex.getErrorCode() == ErrorCodes.FORG0001) {
+                    throw new XPathException(ErrorCodes.XPTY0004, "cannot apply operator " + operator.generalComparisonSymbol + " to xs:anyURI");
+                }
+                throw ex;
+            }
+
+            return compareTo(collator, operator, atomicValue);
         }
     }
 

--- a/src/org/exist/xquery/value/NumericValue.java
+++ b/src/org/exist/xquery/value/NumericValue.java
@@ -3,6 +3,7 @@ package org.exist.xquery.value;
 import com.ibm.icu.text.Collator;
 import org.exist.xquery.Constants;
 import org.exist.xquery.Constants.Comparison;
+import org.exist.xquery.ErrorCodes;
 import org.exist.xquery.XPathException;
 
 public abstract class NumericValue extends ComputableValue {
@@ -75,7 +76,7 @@ public abstract class NumericValue extends ComputableValue {
                     throw new XPathException("Type error: cannot apply operator to numeric value");
             }
         }
-        throw new XPathException("Type error: cannot compare operands: " +
+        throw new XPathException(ErrorCodes.XPTY0004, "Type error: cannot compare operands: " +
                 Type.getTypeName(getType()) + " and " +
                 Type.getTypeName(other.getType()));
     }

--- a/test/src/xquery/namespaces.xql
+++ b/test/src/xquery/namespaces.xql
@@ -272,3 +272,27 @@ declare
 function nt:qname-lhs-compare() {
     xs:QName('test') eq 'test'
 };
+
+declare
+    %test:assertError("XPTY0004")
+function nt:int-string-compare() {
+    5 eq 'five'
+};
+
+declare
+    %test:assertError("XPTY0004")
+function nt:string-int-compare() {
+    'five' eq 5
+};
+
+declare
+    %test:assertError("XPTY0004")
+function nt:uri-int-compare() {
+    xs:anyURI('5') eq 5
+};
+
+declare
+    %test:assertError("XPTY0004")
+function nt:int-uri-compare() {
+    5 eq xs:anyURI('5')
+};

